lobbyboy/provider.py

```diff
@@ -1,89 +1,148 @@
+import json
+import string
+from abc import ABC, abstractmethod
+from datetime import datetime
 from pathlib import Path
 import logging
 import time
-import threading
+from typing import List, Dict, Callable
 
-from lobbyboy.utils import choose_option
+from paramiko.channel import Channel
+
+from lobbyboy.exceptions import NoAvailableNameException
+from lobbyboy.utils import (
+    confirm_ssh_key_pair,
+    send_to_channel,
+    KeyTypeSupport,
+)
+from lobbyboy.config import LBConfigProvider, LBServerMeta
 
 logger = logging.getLogger(__name__)
+SERVER_FILE = "server.json"
+
 
+class BaseProvider(ABC):
+    def __init__(self, provider_name: str, config: LBConfigProvider, workspace: Path):
+        self.name: str = provider_name
+        self.provider_config: LBConfigProvider = config
+        self.workspace: Path = workspace
 
-class BaseProvider:
-    def __init__(self, provider_name, config, provider_config, data_path):
-        self.config = config
-        self.provider_name = provider_name
-        self.provider_config = provider_config
-        self.data_path = Path(data_path)
+    @staticmethod
+    def time_process_action(c: Channel, act: Callable, max_check: int = 20, interval: int = 3, **action_kws) -> bool:
+        """
+        Args:
+           c: paramiko channel
+           act: execute with time process, need return bool
+           max_check: max check times
+           interval: check interval in seconds
 
-    def new_server(self, channel):
+        Returns:
+            bool: bool result before end of check time
+        """
+        action_name = " ".join(act.__name__.split("_"))
+        send_to_channel(c, f"Check {action_name}", suffix="")
+        start_at = time.time()
+        try_times = 1
+        while try_times <= max_check:
+            send_to_channel(c, ".", suffix="")
+            res = act(**action_kws)
+            if res:
+                send_to_channel(c, f"OK({round(time.time() - start_at, 2)}s).")
+                return res
+            time.sleep(interval)
+            try_times += 1
+        send_to_channel(c, "UNKNOWN state, please check it manually.")
+        return False
+
+    @staticmethod
+    def save_raw_server(server_obj: Dict, server_workspace: Path) -> Path:
+        _path = server_workspace.joinpath(SERVER_FILE)
+        with open(_path, "w+") as f:
+            logger.debug(f"write new server data to {_path}")
+            json.dump(server_obj, f)
+        return _path
+
+    @staticmethod
+    def load_raw_server(server_workspace: Path):
+        _path = server_workspace.joinpath(SERVER_FILE)
+        with open(_path, "r+") as f:
+            logger.debug(f"load server data from {_path}")
+            return json.load(f)
+
+    def _generate_server_workspace(self):
+        server_name = datetime.now().strftime("%Y-%m-%d-%H%M")
```

Comment:
I think we should leave this `server_name` generate to provider.

for e.g. vagrant won't accept `-` in server names.

```
ERR [20211124-11:36:01.322] thr=123145521541120 lobbyboy.contrib.provider.vagrant:70: vagrant_command FAILED! command=vagrant up return_code: 1, stdout: b"Bringing machine 'dev_datadir/vagrant/lobbyboy-2021-11-24-1135' up with 'virtualbox' provider...\n", stdout: b"There are errors in the configuration of this machine. Please fix\nthe following errors and try again:\n\nvm:\n* The sub-VM name 'dev_datadir/vagrant/lobbyboy-2021-11-24-1135' is invalid. Please don't use special characters.\n\n"
CRI [20211124-11:36:01.323] thr=123145521541120 lobbyboy.socket_handle:248: *** Socket thread error.
Traceback (most recent call last):
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/socket_handle.py", line 235, in run
    lb_server, proxy_subprocess = self.prepare_shell_env(server, t)
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/socket_handle.py", line 160, in prepare_shell_env
    proxy_subprocess, lb_server = self._create_proxy_process(server.slave_fd)
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/socket_handle.py", line 95, in _create_proxy_process
    meta: LBServerMeta = self.choose_server()
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/socket_handle.py", line 65, in choose_server
    return self._ask_user_to_create_server()
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/socket_handle.py", line 87, in _ask_user_to_create_server
    meta: LBServerMeta = provider.create_server(self.channel)
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/contrib/provider/vagrant.py", line 30, in create_server
    self.time_process_action(channel, self._run_vagrant, command_exec=["vagrant", "up"], cwd=str(server_workspace))
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/provider.py", line 48, in time_process_action
    res = act(**action_kws)
  File "/Users/xintao.lai/Programs/lobbyboy/lobbyboy/contrib/provider/vagrant.py", line 74, in _run_vagrant
    raise Exception
Exception
```

Reply:
first, the name `_generate_server_workspace` is not good, who knows what i think that time😥
then, i think we should provide a common `server_name` generate rule(of course, it's can be override for every `provider`), generally speaking when i login to do/linode/..., i can see them from the console at a glance.
as last, `server_name` and `server_workspace` is so important to `provider`, maybe we can make this become basic `property` of `provider`(of course it's also be overwritten.)?

Reply:
maybe `server_name` is enough, `server_workspace` = path(provider).joinpath(workspace), like this:

```python
class BaseProvider:

    @property
    def default_server_name(self):
        ...
    
    def get_server_workspace(self, server_name: str) -> Path:
        return self.workspace.joinpath(server_name)
```

in the `provider.py`
```python
class XProvider(BaseProvider):

    def create_server(self, channel: Channel) -> LBServerMeta:
        ...
        server_name = self.default_server_name  # or replace it with provider's rule
        server_workspace = self.get_server_workspace(server_name)
        ...
```

as a word, we only provide some tools on `BaseProvider`, user can use it or implement it by themselves, what do u think?

Reply:
> as a word, we only provide some tools on BaseProvider, user can use it or implement it by themselves, what do u think?

yes, I think the server name should be decided by every provider themself.

as for the helper function, I agree.

but if the `default_server_name` will generate a new one every time, then we better make it a function instead of a property?

Reply:
> but if the `default_server_name` will generate a new one every time, then we better make it a function instead of a property?

agree, let me update😄